File: algorithm-backtracking-implementation-pro/src/n_queens.cpp
#include "backtracking_solver.hpp"
#include <vector>
#include <string>
#include <numeric> // For std::iota, though not strictly needed here

namespace Backtracking {

// Helper function to check if placing a queen at (row, col) is safe
// A position is safe if no other queen attacks it.
// We only need to check queens in previous rows, as we place queens row by row.
bool is_safe(int row, int col, const std::vector<std::string>& board, int n) {
    // Check this column upwards
    for (int i = 0; i < row; ++i) {
        if (board[i][col] == 'Q') {
            return false;
        }
    }

    // Check upper-left diagonal
    for (int i = row - 1, j = col - 1; i >= 0 && j >= 0; --i, --j) {
        if (board[i][j] == 'Q') {
            return false;
        }
    }

    // Check upper-right diagonal
    for (int i = row - 1, j = col + 1; i >= 0 && j < n; --i, ++j) {
        if (board[i][j] == 'Q') {
            return false;
        }
    }

    return true;
}

// Recursive backtracking function to solve N-Queens
// @param row: The current row we are trying to place a queen in
// @param n: The size of the chessboard (N x N)
// @param board: The current state of the board (passed by reference to modify)
// @param results: A vector to store all valid board configurations (solutions)
void backtrack_n_queens(int row, int n,
                        std::vector<std::string>& board,
                        std::vector<std::vector<std::string>>& results) {
    // Base case: If all queens are placed (we have successfully placed a queen in each row)
    if (row == n) {
        results.push_back(board); // Add the current valid configuration to results
        return;
    }

    // Recursive step: Try placing a queen in each column of the current row
    for (int col = 0; col < n; ++col) {
        // Pruning step: Check if it's safe to place a queen at (row, col)
        if (is_safe(row, col, board, n)) {
            // Choice: Place a queen at (row, col)
            board[row][col] = 'Q';

            // Recurse: Move to the next row
            backtrack_n_queens(row + 1, n, board, results);

            // Backtrack: Undo the choice (remove the queen)
            // This is crucial to explore other possibilities in the current row
            board[row][col] = '.';
        }
    }
}

// Main function to solve the N-Queens problem
// @param n: The size of the chessboard
// @return: A vector of vectors of strings, where each inner vector represents a solution
//          and each string in the inner vector represents a row of the board.
std::vector<std::vector<std::string>> solveNQueens(int n) {
    std::vector<std::vector<std::string>> results; // Stores all solutions
    // Initialize an N x N empty board with '.'
    std::vector<std::string> board(n, std::string(n, '.'));

    // Start the backtracking process from the first row (row 0)
    backtrack_n_queens(0, n, board, results);

    return results;
}

} // namespace Backtracking

/*
Complexity Analysis for N-Queens:

Time Complexity:
The theoretical upper bound for N-Queens is O(N!). This is because, in the worst case,
we might explore branches that lead to N! possible permutations of queen placements.
However, due to aggressive pruning by `is_safe`, the actual number of states visited
is significantly less.
A tighter upper bound for the number of solutions is approximately N! / C^N, where C is some constant.
Empirically, the complexity is closer to O(N!). For N=14, there are 365,596 solutions,
but the number of nodes in the search tree is much larger, yet far less than 14! (which is ~8.7 trillion).

Space Complexity:
1. `board`: O(N^2) to store the N x N chessboard.
2. `recursion stack`: The maximum depth of the recursion is N (one call for each row). So, O(N) for the stack space.
3. `results`: In the worst case, there can be many solutions. For N=14, there are 365,596 solutions.
   Each solution takes O(N^2) space. So, the total space for storing results can be O(N_solutions * N^2).
   If we only count the auxiliary space needed by the algorithm excluding the output, it's O(N^2) for the board
   plus O(N) for the recursion stack.

Optimization:
The `is_safe` function can be optimized using boolean arrays or bit manipulation to check for conflicts
in O(1) time. This would typically involve three boolean arrays (or bitmasks):
- `col_used[N]`: to check if a column is occupied.
- `diag1_used[2*N-1]`: to check main diagonals (row + col).
- `diag2_used[2*N-1]`: to check anti-diagonals (row - col + N - 1).
This allows `is_safe` to be an O(1) check, improving the constant factors but not the asymptotic complexity
of the search space exploration. The current implementation of `is_safe` checks linearly, which adds O(N)
to each step. Thus, the current time complexity is more accurately O(N^2 * N!) if counting checks naively,
but since the checks are done on parts of the board, it's generally considered to fall under the N! exploration.
For interview purposes, mentioning the O(1) optimization for `is_safe` is a good point.
*/

File: coding-interview-backtracking-enhanced/src/n_queens.py
"""
Problem: N-Queens

The N-Queens puzzle is the problem of placing N non-attacking queens on an N×N chessboard.
This means no two queens share the same row, column, or diagonal.
Given an integer `n`, return all distinct solutions to the N-Queens puzzle.
Each solution contains a distinct board configuration of the N queens' placement,
where `'Q'` and `'.'` both indicate a queen and an empty space, respectively.

Example 1:
Input: n = 4
Output:
[[".Q..", "...Q", "Q...", "..Q."],
 ["..Q.", "Q...", "...Q", ".Q.."]]

Example 2:
Input: n = 1
Output: [["Q"]]

Constraints:
1 <= n <= 9
"""

from typing import List

def solve_n_queens(n: int) -> List[List[str]]:
    """
    Solves the N-Queens puzzle using backtracking.

    The problem can be framed as placing one queen per row. For each row, we try to
    place a queen in a valid column. If a placement is valid, we move to the next row.
    If no valid column is found for the current row, we backtrack.

    Algorithm:
    1. Initialize `solutions` as an empty list to store all valid board configurations.
    2. Initialize `board` as a list of strings, where each string represents a row
       (e.g., `["." * n]` for an empty board).
       Alternatively, maintain sets for `cols_occupied`, `diag1_occupied`, `diag2_occupied`
       to quickly check for safety.
       - `cols_occupied`: Stores column indices where a queen is placed.
       - `diag1_occupied`: Stores `row - col` values for queens on this diagonal (top-left to bottom-right).
       - `diag2_occupied`: Stores `row + col` values for queens on this diagonal (top-right to bottom-left).
    3. Define a recursive helper function `backtrack(row)`:
       a. Base Case: If `row == n`, it means N queens have been successfully placed.
          Convert the current `board` state into the required string format and add it to `solutions`. Then return.
       b. Recursive Step: Iterate through each column `col` from `0` to `n-1` for the current `row`:
          i. **Check Safety**: Determine if placing a queen at `(row, col)` is safe (i.e., it doesn't attack
             any previously placed queens). This check is efficient using the sets:
             - `col` not in `cols_occupied`
             - `row - col` not in `diag1_occupied`
             - `row + col` not in `diag2_occupied`
          ii. If safe:
              - **Choose**: Place the queen at `(row, col)`:
                - Update `board` (e.g., `board[row][col] = 'Q'`).
                - Add `col`, `row - col`, `row + col` to their respective occupied sets.
              - **Explore**: Recursively call `backtrack(row + 1)` to place the next queen.
              - **Un-choose (Backtrack)**: Remove the queen from `(row, col)`:
                - Reset `board` (e.g., `board[row][col] = '.'`).
                - Remove `col`, `row - col`, `row + col` from their respective occupied sets.
             This step is crucial to explore other possibilities.
    4. Start the backtracking process by calling `backtrack(0)` (starting from the first row).

    Time Complexity: O(N!) - Roughly
        - The upper bound for N-Queens is often cited as O(N!), not N^N. This is because at each step (row),
          we significantly prune the search space. In the first row, there are N choices. In the second, at most N-2 (one column and two diagonals are blocked), etc.
        - The number of actual solutions `S(N)` grows very fast, but is much less than N^N. For N=8, S(8)=92. For N=9, S(9)=352.
        - The recursive calls explore a search tree. The depth of the tree is N. In the worst case, we might visit
          a large portion of the `N!` possible placements if we only check constraints at the end.
          However, with effective pruning (checking safety at each step), the complexity is closer to O(N!)
          because the number of valid partial solutions decreases rapidly.
        - Building a board string from the internal representation for each solution takes O(N^2). So, total
          time complexity can be more accurately described as O(N! * N^2) or even tighter bounds in specific analyses.

    Space Complexity: O(N^2)
        - O(N) for the recursion stack depth.
        - O(N) for the `cols_occupied`, `diag1_occupied`, `diag2_occupied` sets.
        - O(N^2) for the internal `board` representation (e.g., list of lists of chars).
        - The `solutions` list stores `S(N)` boards, each O(N^2) size, so total output space is O(S(N) * N^2).
          Auxiliary space is O(N^2).
    """
    solutions: List[List[str]] = []
    
    # Using sets for O(1) average time complexity for add/remove/check operations
    cols_occupied = set()           # Stores column indices of placed queens
    diag1_occupied = set()          # Stores (row - col) for placed queens (top-left to bottom-right diagonal)
    diag2_occupied = set()          # Stores (row + col) for placed queens (top-right to bottom-left diagonal)

    # `board` will store the current configuration during backtracking.
    # It's a list of strings where board[row] represents the row.
    # Initialize with '.' representing empty squares.
    board = [['.' for _ in range(n)] for _ in range(n)]

    def backtrack(row: int):
        # Base case: All N queens have been successfully placed
        if row == n:
            # Convert the current board (list of lists of chars) to the required
            # list of strings format and add to solutions.
            current_solution = ["".join(row_chars) for row_chars in board]
            solutions.append(current_solution)
            return

        # Iterate through all possible columns in the current row
        for col in range(n):
            # Check if placing a queen at (row, col) is safe
            # A position is safe if no other queen is in the same column,
            # or on either of the two diagonals.
            if (col in cols_occupied or
                (row - col) in diag1_occupied or
                (row + col) in diag2_occupied):
                continue # This position is attacked, try next column

            # If safe:
            # 1. Choose: Place the queen
            board[row][col] = 'Q'
            cols_occupied.add(col)
            diag1_occupied.add(row - col)
            diag2_occupied.add(row + col)

            # 2. Explore: Recurse to place the queen in the next row
            backtrack(row + 1)

            # 3. Un-choose (Backtrack): Remove the queen and revert changes
            # This is essential to explore other placement possibilities for the current row.
            board[row][col] = '.'
            cols_occupied.remove(col)
            diag1_occupied.remove(row - col)
            diag2_occupied.remove(row + col)

    # Start the backtracking process from the first row (row 0)
    backtrack(0)
    return solutions

# Helper function to print a board for visualization (optional)
def print_board(board: List[str]):
    if not board:
        print("Empty board.")
        return
    print("-" * (len(board[0]) + 2))
    for row_str in board:
        print(f"|{row_str}|")
    print("-" * (len(board[0]) + 2))

if __name__ == '__main__':
    # Test cases
    print("N-Queens solutions for n = 4:")
    sols_4 = solve_n_queens(4)
    for i, sol in enumerate(sols_4):
        print(f"Solution {i+1}:")
        print_board(sol)
    # Expected: 2 solutions

    print("\nN-Queens solutions for n = 1:")
    sols_1 = solve_n_queens(1)
    for i, sol in enumerate(sols_1):
        print(f"Solution {i+1}:")
        print_board(sol)
    # Expected: 1 solution

    print("\nN-Queens solutions for n = 2:")
    sols_2 = solve_n_queens(2)
    print(f"Number of solutions: {len(sols_2)}")
    for i, sol in enumerate(sols_2):
        print(f"Solution {i+1}:")
        print_board(sol)
    # Expected: 0 solutions

    print("\nN-Queens solutions for n = 3:")
    sols_3 = solve_n_queens(3)
    print(f"Number of solutions: {len(sols_3)}")
    for i, sol in enumerate(sols_3):
        print(f"Solution {i+1}:")
        print_board(sol)
    # Expected: 0 solutions

    print("\nN-Queens solutions for n = 8:")
    sols_8 = solve_n_queens(8)
    print(f"Number of solutions: {len(sols_8)}")
    # Expected: 92 solutions. (Too many to print here, just count)

    print("\nN-Queens solutions for n = 0 (Edge case):")
    sols_0 = solve_n_queens(0)
    print(f"Number of solutions: {len(sols_0)}")
    print(f"Solutions: {sols_0}")
    # Expected: [[]] (An empty board is technically a valid placement of 0 queens)